### ISSUE TITLE
Add support for multiple mysql instances running of different ports.

### ISSUE THREAD

**Issue by khalidamen** [rackerlabs/recap#136]
Please add support to monitor multiple mysql instances running on different ports.

**Comment by tonyskapunk** [comment]
Hi @khalidamen thanks for the request, some time ago it was added but dropped out due to some concerns about the validation required and not having people using it, but glad to read there's people out there that will use it.

I'm happy to evaluate the implementation and give it a try.

References for this:
- https://github.com/rackerlabs/recap/issues/81#issuecomment-287376750
- https://github.com/rackerlabs/recap/commit/d1d27cdbce58069d099e902fca8c33189ab594a8

**Comment by tonyskapunk** [comment]
#154 implements this feature Here an example

Test env
- Two instances of mysql running
```bash
# ps -eo pid,cmd | grep mysql[d]
17035 /usr/bin/mysqld
18379 mysqld --defaults-file=/etc/mysql/my2.cnf --user=mysql --datadir=/var/lib/mysql2

# ss -npl | grep mysql[d]
u_str     LISTEN      0      80     /run/mysqld/mysqld.sock 5209134     * 0     users:(("mysqld",pid=17035,fd=22))
u_str     LISTEN      0      80    /run/mysqld/mysqld2.sock 5215645     * 0     users:(("mysqld",pid=18379,fd=22))
tcp       LISTEN      0      80                           *:3306        *:*     users:(("mysqld",pid=17035,fd=21))
tcp       LISTEN      0      80                           *:3307        *:*     users:(("mysqld",pid=18379,fd=21))
```

Config
- The first two files exist and can be used on each instance of mysql
- The third file does not exist.
```ini
$ grep ^[^#] /etc/recap.conf 
USEMYSQL="yes"
USEMYSQLPROCESSLIST="yes"
USEINNODB="yes"
DOTMYDOTCNF="/root/.my1.cnf, /root/.my2.cnf, /root/.my.cnf"
```

recap.log
```log
2018-05-04 14:06:18-05:00 [INFO] --- Starting recap[22297] ---
2018-05-04 14:06:18-05:00 [INFO] -- bash info: 4 4 19 1 release x86_64-unknown-linux-gnu
2018-05-04 14:06:18-05:00 [INFO] recap[22297]: Created lock file: /var/lock/recap.lock
2018-05-04 14:06:18-05:00 [INFO] Starting check for disk space
2018-05-04 14:06:18-05:00 [INFO] Ended check for disk space
2018-05-04 14:06:18-05:00 [INFO] -- Report suffix: 20180504-140618
2018-05-04 14:06:18-05:00 [INFO] Starting 'ps' report
2018-05-04 14:06:18-05:00 [INFO] Ended 'ps' report
2018-05-04 14:06:18-05:00 [INFO] Starting 'uptime' report
2018-05-04 14:06:18-05:00 [INFO] Ended 'uptime' report
2018-05-04 14:06:18-05:00 [INFO] Starting 'free' report
2018-05-04 14:06:18-05:00 [INFO] Ended 'free' report
2018-05-04 14:06:18-05:00 [INFO] Starting 'vmstat' report
2018-05-04 14:06:20-05:00 [INFO] Ended 'vmstat' report
2018-05-04 14:06:20-05:00 [INFO] Starting 'iostat' report
2018-05-04 14:06:22-05:00 [INFO] Ended 'iostat' report
2018-05-04 14:06:22-05:00 [INFO] Starting 'iotop' report
2018-05-04 14:06:24-05:00 [INFO] Ended 'iotop' report
2018-05-04 14:06:24-05:00 [INFO] Starting 'sar' report
2018-05-04 14:06:24-05:00 [INFO] Ended 'sar' report
2018-05-04 14:06:24-05:00 [INFO] Starting 'disk utilization' report
2018-05-04 14:06:24-05:00 [INFO] Ended 'disk utilization' report
2018-05-04 14:06:24-05:00 [INFO] Starting 'top 10 cpu' report
2018-05-04 14:06:28-05:00 [INFO] Ended 'top 10 cpu' report
2018-05-04 14:06:28-05:00 [INFO] Starting 'top 10 memory' report
2018-05-04 14:06:28-05:00 [INFO] Ended 'top 10 memory' report
2018-05-04 14:06:28-05:00 [INFO] Starting 'network socket' report
2018-05-04 14:06:28-05:00 [INFO] Ended 'network socket' report
2018-05-04 14:06:28-05:00 [INFO] Starting iteration of mysql report(s)
2018-05-04 14:06:28-05:00 [INFO] Starting 'mysql status' report
2018-05-04 14:06:28-05:00 [INFO] Ended 'mysql status' report
2018-05-04 14:06:28-05:00 [INFO] Starting 'mysql processlist' report
2018-05-04 14:06:28-05:00 [INFO] Ended 'mysql processlist' report
2018-05-04 14:06:28-05:00 [INFO] Starting 'mysql innodb' report
2018-05-04 14:06:29-05:00 [INFO] Ended 'mysql innodb' report
2018-05-04 14:06:29-05:00 [INFO] Starting 'mysql status' report
2018-05-04 14:06:29-05:00 [INFO] Ended 'mysql status' report
2018-05-04 14:06:29-05:00 [INFO] Starting 'mysql processlist' report
2018-05-04 14:06:29-05:00 [INFO] Ended 'mysql processlist' report
2018-05-04 14:06:29-05:00 [INFO] Starting 'mysql innodb' report
2018-05-04 14:06:29-05:00 [INFO] Ended 'mysql innodb' report
2018-05-04 14:06:29-05:00 [ERROR] Unable to read: '/root/.my.cnf' in the 'DOTMYDOTCNF' config.
2018-05-04 14:06:29-05:00 [INFO] Ended iteration of mysql report(s)
2018-05-04 14:06:29-05:00 [INFO] Ended all the reports
2018-05-04 14:06:29-05:00 [INFO] recap[22297]: Caught signal - deleting /var/lock/recap.lock
2018-05-04 14:06:29-05:00 [INFO] Execution time: 11s
2018-05-04 14:06:29-05:00 [INFO] --- Ending recap[22297] ---
```

The mysql report:
```log
# cat /var/log/recap/mysql_20180504-140618.log 
2018-05-04_14:06:18
MySQL (/root/.my1.cnf) status
Uptime: 5230  Threads: 1  Questions: 78  Slow queries: 0  Opens: 18  Flush tables: 1  Open tables: 12  Queries per second avg: 0.014
 
 
MySQL (/root/.my1.cnf) processes
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
| Id | User | Host      | db | Command | Time | State | Info                  | Progress |
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
| 55 | root | localhost |    | Query   | 0    | init  | show full processlist | 0.000    |
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
 
 
MySQL (/root/.my1.cnf) InnoDB status

=====================================
2018-05-04 14:06:28 7f8b91071b00 INNODB MONITOR OUTPUT
=====================================
Per second averages calculated from the last 56 seconds
-----------------
BACKGROUND THREAD
-----------------
srv_master_thread loops: 1 srv_active, 0 srv_shutdown, 5223 srv_idle
srv_master_thread log flush and writes: 5223
----------
SEMAPHORES
----------
OS WAIT ARRAY INFO: reservation count 2
OS WAIT ARRAY INFO: signal count 2
Mutex spin waits 592, rounds 548, OS waits 0
RW-shared spins 2, rounds 60, OS waits 2
RW-excl spins 0, rounds 0, OS waits 0
Spin rounds per wait: 0.93 mutex, 30.00 RW-shared, 0.00 RW-excl
------------
TRANSACTIONS
------------
Trx id counter 2310
Purge done for trx's n:o < 0 undo n:o < 0 state: running but idle
History list length 0
LIST OF TRANSACTIONS FOR EACH SESSION:
---TRANSACTION 0, not started
MySQL thread id 56, OS thread handle 0x7f8b91071b00, query id 83 localhost root init
SHOW ENGINE INNODB STATUS
---TRANSACTION 2307, not started
MySQL thread id 1, OS thread handle 0x7f8b91152b00, query id 0 Waiting for background binlog tasks
--------
FILE I/O
--------
I/O thread 0 state: waiting for completed aio requests (insert buffer thread)
I/O thread 1 state: waiting for completed aio requests (log thread)
I/O thread 2 state: waiting for completed aio requests (read thread)
I/O thread 3 state: waiting for completed aio requests (read thread)
I/O thread 4 state: waiting for completed aio requests (read thread)
I/O thread 5 state: waiting for completed aio requests (read thread)
I/O thread 6 state: waiting for completed aio requests (write thread)
I/O thread 7 state: waiting for completed aio requests (write thread)
I/O thread 8 state: waiting for completed aio requests (write thread)
I/O thread 9 state: waiting for completed aio requests (write thread)
Pending normal aio reads: 0 [0, 0, 0, 0] , aio writes: 0 [0, 0, 0, 0] ,
 ibuf aio reads: 0, log i/o's: 0, sync i/o's: 0
Pending flushes (fsync) log: 0; buffer pool: 0
172 OS file reads, 5 OS file writes, 5 OS fsyncs
0.00 reads/s, 0 avg bytes/read, 0.00 writes/s, 0.00 fsyncs/s
-------------------------------------
INSERT BUFFER AND ADAPTIVE HASH INDEX
-------------------------------------
Ibuf: size 1, free list len 0, seg size 2, 0 merges
merged operations:
 insert 0, delete mark 0, delete 0
discarded operations:
 insert 0, delete mark 0, delete 0
0.00 hash searches/s, 0.00 non-hash searches/s
---
LOG
---
Log sequence number 1616727
Log flushed up to   1616727
Pages flushed up to 1616727
Last checkpoint at  1616727
Max checkpoint age    80826164
Checkpoint age target 78300347
Modified age          0
Checkpoint age        0
0 pending log writes, 0 pending chkp writes
8 log i/o's done, 0.00 log i/o's/second
----------------------
BUFFER POOL AND MEMORY
----------------------
Total memory allocated 140574720; in additional pool allocated 0
Total memory allocated by read views 88
Internal hash tables (constant factor + variable factor)
    Adaptive hash index 2217568 	(2213368 + 4200)
    Page hash           139112 (buffer pool 0 only)
    Dictionary cache    629911 	(554768 + 75143)
    File system         817744 	(812272 + 5472)
    Lock system         333592 	(332872 + 720)
    Recovery system     0 	(0 + 0)
Dictionary memory allocated 75143
Buffer pool size        8191
Buffer pool size, bytes 134201344
Free buffers            8034
Database pages          157
Old database pages      0
Modified db pages       0
Percent of dirty pages(LRU & free pages): 0.000
Max dirty pages percent: 75.000
Pending reads 0
Pending writes: LRU 0, flush list 0, single page 0
Pages made young 0, not young 0
0.00 youngs/s, 0.00 non-youngs/s
Pages read 157, created 0, written 1
0.00 reads/s, 0.00 creates/s, 0.00 writes/s
No buffer pool page gets since the last printout
Pages read ahead 0.00/s, evicted without access 0.00/s, Random read ahead 0.00/s
LRU len: 157, unzip_LRU len: 0
I/O sum[0]:cur[0], unzip sum[0]:cur[0]
--------------
ROW OPERATIONS
--------------
0 queries inside InnoDB, 0 queries in queue
0 read views open inside InnoDB
0 RW transactions active inside InnoDB
0 RO transactions active inside InnoDB
0 out of 1000 descriptors used
Main thread process no. 17035, id 140236845582080, state: sleeping
Number of rows inserted 0, updated 0, deleted 0, read 0
0.00 inserts/s, 0.00 updates/s, 0.00 deletes/s, 0.00 reads/s
Number of system rows inserted 0, updated 0, deleted 0, read 0
0.00 inserts/s, 0.00 updates/s, 0.00 deletes/s, 0.00 reads/s
----------------------------
END OF INNODB MONITOR OUTPUT
============================
MySQL (/root/.my2.cnf) status
Uptime: 3610  Threads: 1  Questions: 39  Slow queries: 0  Opens: 18  Flush tables: 1  Open tables: 12  Queries per second avg: 0.010
 
 
MySQL (/root/.my2.cnf) processes
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
| Id | User | Host      | db | Command | Time | State | Info                  | Progress |
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
| 25 | root | localhost |    | Query   | 0    | init  | show full processlist | 0.000    |
+----+------+-----------+----+---------+------+-------+-----------------------+----------+
 
 
MySQL (/root/.my2.cnf) InnoDB status

=====================================
2018-05-04 14:06:29 7f7e76be7b00 INNODB MONITOR OUTPUT
=====================================
Per second averages calculated from the last 56 seconds
-----------------
BACKGROUND THREAD
-----------------
srv_master_thread loops: 1 srv_active, 0 srv_shutdown, 3604 srv_idle
srv_master_thread log flush and writes: 3605
----------
SEMAPHORES
----------
OS WAIT ARRAY INFO: reservation count 3
OS WAIT ARRAY INFO: signal count 3
Mutex spin waits 542, rounds 456, OS waits 1
RW-shared spins 2, rounds 60, OS waits 2
RW-excl spins 0, rounds 0, OS waits 0
Spin rounds per wait: 0.84 mutex, 30.00 RW-shared, 0.00 RW-excl
------------
TRANSACTIONS
------------
Trx id counter 3846
Purge done for trx's n:o < 0 undo n:o < 0 state: running but idle
History list length 0
LIST OF TRANSACTIONS FOR EACH SESSION:
---TRANSACTION 0, not started
MySQL thread id 26, OS thread handle 0x7f7e76be7b00, query id 44 localhost root init
SHOW ENGINE INNODB STATUS
---TRANSACTION 3843, not started
MySQL thread id 1, OS thread handle 0x7f7e76cc8b00, query id 0 Waiting for background binlog tasks
--------
FILE I/O
--------
I/O thread 0 state: waiting for completed aio requests (insert buffer thread)
I/O thread 1 state: waiting for completed aio requests (log thread)
I/O thread 2 state: waiting for completed aio requests (read thread)
I/O thread 3 state: waiting for completed aio requests (read thread)
I/O thread 4 state: waiting for completed aio requests (read thread)
I/O thread 5 state: waiting for completed aio requests (read thread)
I/O thread 6 state: waiting for completed aio requests (write thread)
I/O thread 7 state: waiting for completed aio requests (write thread)
I/O thread 8 state: waiting for completed aio requests (write thread)
I/O thread 9 state: waiting for completed aio requests (write thread)
Pending normal aio reads: 0 [0, 0, 0, 0] , aio writes: 0 [0, 0, 0, 0] ,
 ibuf aio reads: 0, log i/o's: 0, sync i/o's: 0
Pending flushes (fsync) log: 0; buffer pool: 0
172 OS file reads, 5 OS file writes, 5 OS fsyncs
0.00 reads/s, 0 avg bytes/read, 0.00 writes/s, 0.00 fsyncs/s
-------------------------------------
INSERT BUFFER AND ADAPTIVE HASH INDEX
-------------------------------------
Ibuf: size 1, free list len 0, seg size 2, 0 merges
merged operations:
 insert 0, delete mark 0, delete 0
discarded operations:
 insert 0, delete mark 0, delete 0
0.00 hash searches/s, 0.00 non-hash searches/s
---
LOG
---
Log sequence number 1616757
Log flushed up to   1616757
Pages flushed up to 1616757
Last checkpoint at  1616757
Max checkpoint age    80826164
Checkpoint age target 78300347
Modified age          0
Checkpoint age        0
0 pending log writes, 0 pending chkp writes
8 log i/o's done, 0.00 log i/o's/second
----------------------
BUFFER POOL AND MEMORY
----------------------
Total memory allocated 140574720; in additional pool allocated 0
Total memory allocated by read views 88
Internal hash tables (constant factor + variable factor)
    Adaptive hash index 2217568 	(2213368 + 4200)
    Page hash           139112 (buffer pool 0 only)
    Dictionary cache    629911 	(554768 + 75143)
    File system         817744 	(812272 + 5472)
    Lock system         333592 	(332872 + 720)
    Recovery system     0 	(0 + 0)
Dictionary memory allocated 75143
Buffer pool size        8191
Buffer pool size, bytes 134201344
Free buffers            8034
Database pages          157
Old database pages      0
Modified db pages       0
Percent of dirty pages(LRU & free pages): 0.000
Max dirty pages percent: 75.000
Pending reads 0
Pending writes: LRU 0, flush list 0, single page 0
Pages made young 0, not young 0
0.00 youngs/s, 0.00 non-youngs/s
Pages read 157, created 0, written 1
0.00 reads/s, 0.00 creates/s, 0.00 writes/s
No buffer pool page gets since the last printout
Pages read ahead 0.00/s, evicted without access 0.00/s, Random read ahead 0.00/s
LRU len: 157, unzip_LRU len: 0
I/O sum[0]:cur[0], unzip sum[0]:cur[0]
--------------
ROW OPERATIONS
--------------
0 queries inside InnoDB, 0 queries in queue
0 read views open inside InnoDB
0 RW transactions active inside InnoDB
0 RO transactions active inside InnoDB
0 out of 1000 descriptors used
Main thread process no. 18379, id 140180567353088, state: sleeping
Number of rows inserted 0, updated 0, deleted 0, read 0
0.00 inserts/s, 0.00 updates/s, 0.00 deletes/s, 0.00 reads/s
Number of system rows inserted 0, updated 0, deleted 0, read 0
0.00 inserts/s, 0.00 updates/s, 0.00 deletes/s, 0.00 reads/s
----------------------------
END OF INNODB MONITOR OUTPUT
============================
```

**Comment by tonyskapunk** [comment]
Sorry click the wrong button.

@khalidamen could you please test this one?

**Comment by tonyskapunk** [comment]
@khalidamen Any chance on testing this?

**Comment by tonyskapunk** [comment]
I added another commit, I included timeouts in mysql commands and validate connectivity prior running any mysql report.

Here the steps to create a testing env:

```bash
## Using mariadb and creating 2 instances from scratch

### Create extra directory
sudo mkdir /var/lib/mysql2

### Create data
sudo mysql_install_db --user=mysql --basedir=/usr --datadir=/var/lib/mysql
sudo mysql_install_db --user=mysql --basedir=/usr --datadir=/var/lib/mysql2


### Create 2nd instance config:
sudo cp -f /etc/mysql/my{,2}.cnf
sudo sed -i -e 's/mysqld.sock/mysqld2.sock/g' -e 's/3306/3307/' /etc/mysql/my2.cnf

### Start the instances
sudo mysqld --defaults-file=/etc/mysql/my.cnf --user=mysql --datadir=/var/lib/mysql &
sudo mysqld --defaults-file=/etc/mysql/my2.cnf --user=mysql --datadir=/var/lib/mysql2 &

### Change passwords through secure install

#### Use aTWvbUwejyh4dE
sudo mysql_secure_installation --defaults-file=/etc/mysql/my.cnf

#### Use Vro9mRnd9Ne9v9
sudo mysql_secure_installation --defaults-file=/etc/mysql/my2.cnf

### Create client config files

sudo tee /root/.my1.cnf << EOF
[client]
socket = /run/mysqld/mysqld.sock
password = aTWvbUwejyh4dE
EOF

sudo tee /root/.my2.cnf << EOF
[client]
socket = /run/mysqld/mysqld2.sock
password = Vro9mRnd9Ne9v9
EOF


### Test connectivity to both instances:

sudo mysqladmin --defaults-file=/root/.my1.cnf ping 
mysqld is alive

sudo mysqladmin --defaults-file=/root/.my2.cnf ping 
mysqld is alive


### With recap installed, configure to define multiple config files:

sudo tee -a /etc/recap.conf <<EOF
USEMYSQL="yes"
USEMYSQLPROCESSLIST="yes"
USEINNODB="yes"
DOTMYDOTCNF="/root/.my1.cnf,/root/.my2.cnf, /root/.my.cnf"
EOF


### Run recap to confirm it works as expected, the `/root/.my.cnf` does not exist, logs should reflect this

sudo recap
sudo tail /var/log/recap/recap.log | grep ERR
2018-06-22 10:59:32-05:00 [ERROR] Unable to read: '/root/.my.cnf' in the 'DOTMYDOTCNF' config.

### Stop one of the instances and re-run recap

sudo kill -STOP $(pidof mysqld|awk '{print $1}')
sudo recap

sudo tail /var/log/recap/recap.log|grep ERR
2018-06-22 11:47:55-05:00 [ERROR] Unable to connect using '/root/.my2.cnf' in the 'DOTMYDOTCNF' config.
2018-06-22 11:47:55-05:00 [ERROR] Unable to read: '/root/.my.cnf' in the 'DOTMYDOTCNF' config.

```

**Comment by tonyskapunk** [comment]
The code that implements this feature has been merged (#154 ) in `development` closing this one.
